packages/flutter/lib/src/rendering/sliver_multi_box_adaptor.dart

```diff
@@ -224,23 +250,75 @@ abstract class RenderSliverMultiBoxAdaptor extends RenderSliver
 
   bool _debugAssertChildListLocked() => childManager.debugAssertChildListLocked();
 
+  /// Verify the child list index is in strictly increasing order.
+  ///
+  /// This is a no-op for non-debug builds.
+  void _debugVerifyChildOrder(){
+    assert(() {
+      if (_debugChildIntegrityEnabled) {
+        if (firstChild != null) {
+          int index = indexOf(firstChild);
+          RenderBox child = childAfter(firstChild);
+          while (child != null) {
+            assert(indexOf(child) > index);
+            index = indexOf(child);
+            child = childAfter(child);
+          }
+        }
+      }
+      return true;
+    }());
+  }
+
   @override
   void insert(RenderBox child, { RenderBox after }) {
     assert(!_keepAliveBucket.containsValue(child));
     super.insert(child, after: after);
     assert(firstChild != null);
     assert(() {
-      int index = indexOf(firstChild);
-      RenderBox child = childAfter(firstChild);
-      while (child != null) {
-        assert(indexOf(child) > index);
-        index = indexOf(child);
-        child = childAfter(child);
-      }
+      _debugVerifyChildOrder();
       return true;
     }());
   }
 
+  @override
+  void move(RenderBox child, { RenderBox after }) {
+    // There are two scenarios:
+    //
+    // 1, The child is not keptAlive.
+    // The child is in childList maintained by ContainerRenderObjectMixin.
+    // We can call the super.move and update parentData with new slot.
+    //
+    // 2, The child is keptAlive.
+    // In this case, the child is no longer in the childList but might be stored in
+    // [_keepAliveBucket]. We need to update the location of the child in the bucket.
+    final SliverMultiBoxAdaptorParentData childParentData = child.parentData;
+    if (!childParentData.keptAlive) {
+      super.move(child, after: after);
+      // The following call updates the slot in the parentData.
+      childManager.didAdoptChild(child);
+    } else {
+      // If the child in the bucket is not current child, that means someone has
+      if (_keepAliveBucket[childParentData.index] == child) {
+        _keepAliveBucket.remove(childParentData.index);
+      }
+      assert(() {
+        _debugDanglingKeepAlives.remove(child);
+        return true;
+      }());
+      childManager.didAdoptChild(child);
+      // If there is an existing child in the new slot, that mean that child will
+      // be moved to other index. In other cases, the existing child should have been
+      // removed by updateChild. Thus, it is ok to overwrite it.
+      assert(() {
+        if (_keepAliveBucket.containsKey(childParentData.index))
+          _debugDanglingKeepAlives.add(_keepAliveBucket[childParentData.index]);
```

Comment:
Do we need to remove things from _debugDanglingKeepAlives if `remove` below is called? If yes, please make sure you have a failing test if this doesn't happen.

Reply:
no, this is a tricky one. If there is a dangling child, it will have to be swapped in the later move. In other cases, the dangling child should have been removed in line 1106 in sliver.dart.

Reply:
Hm, but that's only true because of how things are ordered in sliver.dart, no? It's kinda brittle to rely in that implementation detail here...

Reply:
Added in the remove function